File: android/app/src/main/java/io/tastecard/engine/TextEmbeddingStore.kt
package io.tastecard.engine

import android.content.Context
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Loads the precomputed, L2-normalised category text vectors (§6), produced by
 * scripts/precompute_text_embeddings.py. Same TCTE binary format as iOS:
 *   "TCTE" | u32 version | u32 dim | u32 count | { u16 idLen, id bytes, dim*f32 }...
 */
class TextEmbeddingStore private constructor(
    val dimension: Int,
    private val vectors: Map<String, FloatArray>,
) {
    fun vector(categoryId: String): FloatArray? = vectors[categoryId]

    class MissingModelException(message: String) : Exception(message)

    companion object {
        fun loadFromAssets(context: Context, name: String = "category_text_embeddings.bin"): TextEmbeddingStore {
            val bytes = try {
                context.assets.open(name).use { it.readBytes() }
            } catch (e: Exception) {
                throw MissingModelException("text vectors not bundled: ${e.message}")
            }
            return parse(bytes)
        }

        fun parse(bytes: ByteArray): TextEmbeddingStore {
            val buf = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
            val magic = ByteArray(4); buf.get(magic)
            if (String(magic, Charsets.US_ASCII) != "TCTE") throw MissingModelException("bad magic")
            val version = buf.int
            if (version != 1) throw MissingModelException("unsupported version $version")
            val dim = buf.int
            val count = buf.int
            if (dim <= 0 || count <= 0) throw MissingModelException("empty text vectors")

            val map = HashMap<String, FloatArray>(count)
            repeat(count) {
                val idLen = buf.short.toInt() and 0xFFFF
                val idBytes = ByteArray(idLen); buf.get(idBytes)
                val id = String(idBytes, Charsets.UTF_8)
                val vec = FloatArray(dim) { buf.float }
                map[id] = vec
            }
            return TextEmbeddingStore(dim, map)
        }
    }
}

File: Tastecard/Engine/TextEmbeddingStore.swift
//
//  TextEmbeddingStore.swift
//  Tastecard
//
//  Loads the precomputed, prompt-ensembled, L2-normalised category text embeddings
//  produced at build time by scripts/precompute_text_embeddings.py (§6). The text
//  encoder NEVER ships — only these vectors do.
//
//  Binary format (little-endian), self-describing so order/version are explicit:
//      magic   : 4 bytes  = "TCTE"
//      version : UInt32   = 1
//      dim     : UInt32
//      count   : UInt32
//      repeated count times:
//          idLen : UInt16
//          id    : idLen UTF-8 bytes
//          vec   : dim × Float32   (already L2-normalised)
//

import Foundation

struct TextEmbeddingStore {
    let dimension: Int
    private let vectors: [String: [Float]]

    var categoryIds: [String] { Array(vectors.keys) }
    func vector(for categoryId: String) -> [Float]? { vectors[categoryId] }

    static func loadBundled(bundle: Bundle = .main) throws -> TextEmbeddingStore {
        guard let url = bundle.url(forResource: "category_text_embeddings", withExtension: "bin") else {
            throw EmbedderError.modelMissing
        }
        let data = try Data(contentsOf: url)
        return try parse(data)
    }

    static func parse(_ data: Data) throws -> TextEmbeddingStore {
        var cursor = 0

        func read<T>(_ type: T.Type) throws -> T {
            let size = MemoryLayout<T>.size
            guard cursor + size <= data.count else {
                throw EmbedderError.modelLoadFailed("text embeddings truncated")
            }
            let value = data.subdata(in: cursor..<cursor + size).withUnsafeBytes {
                $0.loadUnaligned(as: T.self)
            }
            cursor += size
            return value
        }

        func readBytes(_ n: Int) throws -> Data {
            guard cursor + n <= data.count else {
                throw EmbedderError.modelLoadFailed("text embeddings truncated")
            }
            let slice = data.subdata(in: cursor..<cursor + n)
            cursor += n
            return slice
        }

        let magic = try readBytes(4)
        guard magic == Data("TCTE".utf8) else {
            throw EmbedderError.modelLoadFailed("bad magic in text embeddings")
        }
        let version: UInt32 = try read(UInt32.self)
        guard version == 1 else {
            throw EmbedderError.modelLoadFailed("unsupported text-embedding version \(version)")
        }
        let dim = Int(try read(UInt32.self))
        let count = Int(try read(UInt32.self))
        guard dim > 0, count > 0 else {
            throw EmbedderError.modelLoadFailed("empty text embeddings")
        }

        var map: [String: [Float]] = [:]
        map.reserveCapacity(count)
        for _ in 0..<count {
            let idLen = Int(try read(UInt16.self))
            let idData = try readBytes(idLen)
            guard let id = String(data: idData, encoding: .utf8) else {
                throw EmbedderError.modelLoadFailed("bad category id encoding")
            }
            let vecBytes = try readBytes(dim * MemoryLayout<Float32>.size)
            let vec: [Float] = vecBytes.withUnsafeBytes { raw in
                Array(raw.bindMemory(to: Float32.self))
            }
            map[id] = vec
        }
        return TextEmbeddingStore(dimension: dim, vectors: map)
    }
}
